src/realm/alloc_slab.cpp

```diff
@@ -1242,6 +1274,545 @@ void SlabAlloc::update_reader_view(size_t file_size)
     // scheduling queue.
     //
     rebuild_translations(replace_last_mapping, old_num_mappings);
+
+    if (db) {
+        db->refresh_encrypted_mappings(version, *this);
+    }
+}
+
+static inline ref_type get_ref_from_array(Array& array, size_t index)
+{
+    int64_t val = array.get(index);
+    REALM_ASSERT_EX(!int_cast_has_overflow<ref_type>(val), val, index, array.get_ref());
+    return ref_type(val);
+}
+
+void SlabAlloc::refresh_pages_for_versions(std::vector<VersionedTopRef> read_locks,
```

Comment:
I think we actually don't need this function at all? The encryption layer can determine on its own which pages are dirty because the IV on disk will not match the IV it used when reading that page, so we don't have to reconstruct that information from the changes to the freelist.

Reply:
I should add an explanation comment like this: 
```
// Determine which pages have been modified when advancing versions by comparing
// the freelists to compute where allocations have been made. All intermediate versions
// must be present because allocations from an intermediate version may be freed by a
// later version in the sequence which would show up as no net modification if only
// comparing the first and final version.
//
// When advancing from version A to B:
//
// versions = get versions A through B from the DB.
//
// for each version V in versions {
//    if not version A {
//        get topref, mark page it points to for refresh (possibly next page also)
//        read top array
//        get refs for freelist positions and sizes arrays and mark them for refresh
//        create accessor for those arrays
//        compare freelist to previous freelist, determine which allocations have been made
//        for each allocation, mark relevant pages for refresh
//    } else {
//        previous freelist does not need refresh.
//    }
//    previous freelist = current freelist
// }
```

Reply:
> The encryption layer can determine on its own which pages are dirty because the IV on disk will not match the IV it used when reading that page

This is an interesting idea. But wouldn't we end up refreshing a lot of extra pages unnecessarily?

Reply:
Why would that ever result in an unnecessary refresh?

Reply:
I might not quite understand the idea. But say we are advancing from version 1 to 2 which only requires refreshing page A, but the file has 5 versions written and has modified pages A and B in the later versions. If we check all pages for mismatching IVs when advancing from 1->2, this will fetch page B unnecessarily. This wouldn't be a problem, except if another write comes in to modify page B by the time of the next version advance, then we'd need to fetch B again, and since nothing was ever read from B there was an unnecessary refresh.

Reply:
>  But say we are advancing from version 1 to 2 which only requires refreshing page A, but the file has 5 versions written

That isn't a legal thing to do. You can only advance to a version other than the latest if you already have a transaction at that version, and if you have a transaction at that version then advancing doesn't require refreshing anything.

Reply:
That sounds like a good design, and its too bad we didn't think of it earlier. This direction is worth exploring, especially if it removes the requirement of having `DB::m_last_encryption_page_reader` as a full reader lock because that enables intermediate version cleanup with encryption as well. I will try and build it out in a different branch.

Reply:
As I understand it this will necessitate scanning all IV blocks in the file on each refresh point in order to detect changes. Is this a reasonable cost for handling small changes?

Reply:
@tgoyne wrt to changing the allocator to only use free pages - I have a branch which does just that and performance degenerates completely due to fragmentation. It might be an option in the future, but it will require much more aggressive online defragmentation.

Reply:
Rereading the IV tables could be done lazily. Advancing to a new version would mark all IV tables as stale. Inside the read barrier, we check if the relevant IV page is stale. If so, reread that one IV page and compare, marking pages with changed IVs as stale. After that, check if page is stale and reread if so.

Without doing this we'd be reading 1/65 of the file for 4k pages and 1/257 for 16k.

Reply:
I have made this optimization in https://github.com/realm/realm-core/pull/6304
I will make some performance measurements soon.